contrib/devtools/test_deterministic_coverage.sh

```diff
@@ -12,6 +12,9 @@ export LC_ALL=C
 # Use GCOV_EXECUTABLE="llvm-cov gcov" if compiling with clang.
 GCOV_EXECUTABLE="gcov"
 
+# Keep random seed constant over multiple runs.
+export RANDOM_CTX_SEED=${RANDOM_CTX_SEED:-$(xxd -l 32 -p /dev/urandom | tr -d " \n")}
```

Comment:
Huh? If you change the seed between runs and then compare the results of those runs, you are pretty much guaranteed to get non-deterministic coverage. I think this test should check for non-determinism across runs, assuming everything kept constant. That is, it should act as a proxy to be able to answer the question "How reliable can test failures be reproduced".

Reply:
I think you are looking for the test to answer the question "How deterministic is coverage across runs", which is also a useful question to ask. Not sure how to accommodate both.

Reply:
And a random seed is already picked per process in the while loop you referenced below. So moving the `export RANDOM_CTX_SEED` wouldn't make any observable difference.


So either I leave it as is or remove the `export RANDOM_CTX_SEED` from this file again. What do you think?

Reply:
Sorry for the delay,

> Huh? If you change the seed between runs and then compare the results of those runs, you are pretty much guaranteed to get non-deterministic coverage. I think this test should check for non-determinism across runs, assuming everything kept constant.

You're right, many tests fail with a fixed seed (I tried 5 different tests from the list in the script and all of them exhibited non-deterministic coverage after a few tens of runs), thanks for pointing this out! On the other hand, I've been running the script for all tests which are not in the list inside the script while having the seed changed every time (I moved the line inside the loop just for this test) and no non-determinism in coverage has been detected so far (I'm at iteration ~70 now). I'll update when it gets to a 1000.

> ...it should act as a proxy to be able to answer the question "How reliable can test failures be reproduced"

But then I think it doesn't perfectly serve it's purpose - the fact is that the script's actual mechanism of operation is to compare the coverage between runs and look for discrepancies. If the coverage is found to be the same between runs, it still does not guarantee that a failure can always be reproduced. That's because for a test that might fail, at some point in the code there is a branching into two outcomes - test failing and test succeeding. Since all tests probably always succeed when running this script, the succeeding branch is always taken and we never know what's going on in the failing branch. There might be non-determinism there which could still make the failure irreproducible.

> And a random seed is already picked per process in the while loop you referenced below. So moving the `export RANDOM_CTX_SEED` wouldn't make any observable difference.
> 
> So either I leave it as is or remove the `export RANDOM_CTX_SEED` from this file again. What do you think?

Isn't the seed set to RANDOM_CTX_SEED if this environment variable is found and thus the seed is the same for all runs in the loop if this line is located outside the loop?

Reply:
> Isn't the seed set to RANDOM_CTX_SEED if this environment variable is found and thus the seed is the same for all runs in the loop if this line is located outside the loop?

Yes, but putting it *inside* the loop will make it pick a new seed every time. This is already the default behaviour when no seed is provided at all. So might as well leave it out completely.

Reply:
Oh ok, so yes, I think it's best to remove is then.

Reply:
Ok, removed. Thanks for the discussion!

Reply:
Likewise!

Reply:
BTW,
> ... I'll update when it gets to a 1000.

The script found non-deterministic behaviour at iteration 399, meaning it ran the entire test suite 398 times with a different seed each run and did not detect non-deterministic coverage in any of them. The non-determinism in run 399 might be caused by a test that should be included in the list of tests in the script. I'm not sure how to isolate the test that causes this non-determinism though. It requires some detective work. I'll try working on it.